blob/src/main/java/com/bazaarvoice/emodb/blob/db/s3/S3StorageProvider.java

```diff
@@ -0,0 +1,399 @@
+import com.amazonaws.services.s3.model.InitiateMultipartUploadResult;
+import com.amazonaws.services.s3.model.ListObjectsV2Request;
+import com.amazonaws.services.s3.model.ListObjectsV2Result;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.PartETag;
+import com.amazonaws.services.s3.model.PutObjectRequest;
+import com.amazonaws.services.s3.model.S3Object;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import com.amazonaws.services.s3.model.UploadPartRequest;
+import com.amazonaws.services.s3.model.UploadPartResult;
+import com.amazonaws.util.IOUtils;
+import com.bazaarvoice.emodb.blob.api.Range;
+import com.bazaarvoice.emodb.blob.api.StreamSupplier;
+import com.bazaarvoice.emodb.blob.core.BlobPlacement;
+import com.bazaarvoice.emodb.table.db.astyanax.PlacementFactory;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.io.ByteStreams;
+import com.netflix.astyanax.connectionpool.exceptions.ConnectionException;
+import org.apache.commons.codec.binary.Hex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import javax.ws.rs.core.Response;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.security.DigestInputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class S3StorageProvider {
+    private static final Logger LOGGER = LoggerFactory.getLogger(S3StorageProvider.class);
+
+    private static final int MAX_SCAN_BATCH_SIZE = 100;
+    private static final int DEFAULT_READ_RANGE_LENGTH = 128 * 1024;     //128kb
+    private static final int DEFAULT_WRITE_CHUNK_SIZE = 5 * 1024 * 1024; //5mb
+
+    private final PlacementFactory _placementFactory;
+    private final LoadingCache<String, BlobPlacement> _placementCache;
+
+    @Inject
+    public S3StorageProvider(final PlacementFactory placementFactory) {
+        _placementFactory = Objects.requireNonNull(placementFactory);
+        _placementCache = CacheBuilder.newBuilder().build(new CacheLoader<String, BlobPlacement>() {
+            @Override
+            public BlobPlacement load(final String placement) throws ConnectionException {
+                return (BlobPlacement) _placementFactory.newPlacement(placement);
+            }
+        });
+    }
+
+    private AmazonS3URI getAmazonS3URI(final String tableName, final String tablePlacement, @Nullable final String blobId) {
```

Comment:
At first glance, I like that we kept it simple and straightforward with data organization and naming here (as opposed the dark magic Emo does with Cassandra). That being said, can we ensure that putting the entirety of a single table under the same prefix won't create hotspots in S3?

Reply:
according to https://docs.aws.amazon.com/AmazonS3/latest/dev/optimizing-performance.html we should be fine.